File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    TreeNode ancestor = null;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        bfs(root, p, q);
        
        return this.ancestor;

    }

    private TreeNode bfs(TreeNode parent, TreeNode p, TreeNode q){
        
        if(parent == null)
            return null;
        
        
        TreeNode leftChild = bfs(parent.left, p, q );
        TreeNode rightChild = bfs(parent.right, p , q);
        
        if(leftChild != null && rightChild != null){
            this.ancestor = parent;
            return null;
        }else if(leftChild != null && rightChild == null){
            if(parent.val == p.val || parent.val == q.val){
                this.ancestor = parent;
                return null;
            }else{
                return leftChild;
            }
        }else if(leftChild == null && rightChild != null){
            if(parent.val == p.val || parent.val == q.val){
                this.ancestor = parent;
                return null;
            }else{
                return rightChild;
            }
        }else{
            if(parent.val == p.val || parent.val == q.val){
                return parent;
            }else{
                return null;
            }
        }
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def __init__(self):
        self.ans = True
    
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        leftLen = self.longestPathChild(root.left)
        rightLen = self.longestPathChild(root.right)
        if abs(rightLen - leftLen) >= 2:
            self.ans = False
        return self.ans    

    def longestPathChild(self, node: Optional[TreeNode]) ->int :
        if not node:
            return 0

        leftLen = self.longestPathChild(node.left)
        rightLen = self.longestPathChild(node.right)
        
        if abs(rightLen - leftLen) >= 2:
            self.ans = False

        return max(leftLen, rightLen) + 1    
        